// @vssue/api-github-v3/lib/index.js
import * as tslib_1 from "tslib";
import axios from 'axios';
import { buildURL, concatURL, getCleanURL, parseQuery, } from '@vssue/utils';
import { normalizeUser, normalizeIssue, normalizeComment, normalizeReactions, mapReactionName, } from './utils';
/**
 * Github REST API v3
 *
 * @see https://developer.github.com/v3/
 * @see https://developer.github.com/apps/building-oauth-apps/
 */
var GithubV3 = /** @class */ (function () {
    function GithubV3(_a) {
        var _b = _a.baseURL, baseURL = _b === void 0 ? 'https://github.com' : _b, owner = _a.owner, repo = _a.repo, labels = _a.labels, clientId = _a.clientId, clientSecret = _a.clientSecret, state = _a.state;
        this.baseURL = baseURL;
        this.owner = owner;
        this.repo = repo;
        this.labels = labels;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.state = state;
        this.$http = axios.create({
            baseURL: baseURL === 'https://github.com' ? 'https://api.github.com' : concatURL(baseURL, 'api/v3'),
            headers: {
                'Accept': 'application/vnd.github.v3+json',
            },
        });
        this.$http.interceptors.response.use(function (response) {
            if (response.data.error) {
                return Promise.reject(response.data.error_description);
            }
            return response;
        });
    }
    Object.defineProperty(GithubV3.prototype, "platform", {
        /**
         * The platform api info
         */
        get: function () {
            return {
                name: 'GitHub',
                link: this.baseURL,
                version: 'v3',
                meta: {
                    reactable: true,
                    sortable: false,
                },
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Redirect to the authorization page of platform.
     *
     * @see https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/#1-request-a-users-github-identity
     */
    GithubV3.prototype.redirectAuth = function () {
        window.location.href = buildURL(concatURL(this.baseURL, 'login/oauth/authorize'), {
            client_id: this.clientId,
            redirect_uri: window.location.href,
            scope: 'public_repo',
            state: this.state,
        });
    };
    /**
     * Handle authorization.
     *
     * @return A string for access token, `null` for no authorization code
     *
     * @see https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/
     *
     * @remarks
     * If the `code` and `state` exist in the query, and the `state` matches, remove them from query, and try to get the access token.
     */
    GithubV3.prototype.handleAuth = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, code, replaceURL, accessToken;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = parseQuery(window.location.search);
                        if (!query.code) return [3 /*break*/, 2];
                        if (query.state !== this.state) {
                            return [2 /*return*/, null];
                        }
                        code = query.code;
                        delete query.code;
                        delete query.state;
                        replaceURL = buildURL(getCleanURL(window.location.href), query) + window.location.hash;
                        window.history.replaceState(null, '', replaceURL);
                        return [4 /*yield*/, this.getAccessToken({ code: code })];
                    case 1:
                        accessToken = _a.sent();
                        return [2 /*return*/, accessToken];
                    case 2: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Get user access token via `code`
     *
     * @param options.code - The code from the query
     *
     * @return User access token
     *
     * @see https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/#2-users-are-redirected-back-to-your-site-by-github
     */
    GithubV3.prototype.getAccessToken = function (_a) {
        var code = _a.code;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.post("https://cors-anywhere.herokuapp.com/" + concatURL(this.baseURL, 'login/oauth/access_token'), {
                            client_id: this.clientId,
                            client_secret: this.clientSecret,
                            code: code,
                        }, {
                            headers: {
                                'Accept': 'application/json',
                            },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, data.access_token];
                }
            });
        });
    };
    /**
     * Get the logined user with access token.
     *
     * @param options.accessToken - User access token
     *
     * @return The user
     *
     * @see https://developer.github.com/v3/users/#get-the-authenticated-user
     */
    GithubV3.prototype.getUser = function (_a) {
        var accessToken = _a.accessToken;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.get('user', {
                            headers: { 'Authorization': "token " + accessToken },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeUser(data)];
                }
            });
        });
    };
    /**
     * Get issue of this page according to the issue id or the issue title
     *
     * @param options.accessToken - User access token
     * @param options.issueId - The id of issue
     * @param options.issueTitle - The title of issue
     *
     * @return The raw response of issue
     *
     * @see https://developer.github.com/v3/issues/#list-issues-for-a-repository
     * @see https://developer.github.com/v3/issues/#get-a-single-issue
     * @see https://developer.github.com/v3/#pagination
     */
    GithubV3.prototype.getIssue = function (_a) {
        var accessToken = _a.accessToken, issueId = _a.issueId, issueTitle = _a.issueTitle;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var options, data, e_1, data, issue;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = {};
                        if (accessToken) {
                            options.headers = {
                                'Authorization': "token " + accessToken,
                            };
                        }
                        if (!issueId) return [3 /*break*/, 5];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        options.params = {
                            // to avoid caching
                            timestamp: Date.now(),
                        };
                        return [4 /*yield*/, this.$http.get("repos/" + this.owner + "/" + this.repo + "/issues/" + issueId, options)];
                    case 2:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeIssue(data)];
                    case 3:
                        e_1 = _b.sent();
                        if (e_1.response && e_1.response.status === 404) {
                            return [2 /*return*/, null];
                        }
                        else {
                            throw e_1;
                        }
                        return [3 /*break*/, 4];
                    case 4: return [3 /*break*/, 7];
                    case 5:
                        options.params = {
                            labels: this.labels.join(','),
                            sort: 'created',
                            direction: 'asc',
                            // to avoid caching
                            timestamp: Date.now(),
                        };
                        return [4 /*yield*/, this.$http.get("repos/" + this.owner + "/" + this.repo + "/issues", options)];
                    case 6:
                        data = (_b.sent()).data;
                        issue = data.map(normalizeIssue).find(function (item) { return item.title === issueTitle; });
                        return [2 /*return*/, issue || null];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Create a new issue
     *
     * @param options.accessToken - User access token
     * @param options.title - The title of issue
     * @param options.content - The content of issue
     *
     * @return The created issue
     *
     * @see https://developer.github.com/v3/issues/#create-an-issue
     */
    GithubV3.prototype.postIssue = function (_a) {
        var accessToken = _a.accessToken, title = _a.title, content = _a.content;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.post("repos/" + this.owner + "/" + this.repo + "/issues", {
                            title: title,
                            body: content,
                            labels: this.labels,
                        }, {
                            headers: { 'Authorization': "token " + accessToken },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeIssue(data)];
                }
            });
        });
    };
    /**
     * Get comments of this page according to the issue id or the issue title
     *
     * @param options.accessToken - User access token
     * @param options.issueId - The id of issue
     * @param options.query - The query parameters
     *
     * @return The comments
     *
     * @see https://developer.github.com/v3/issues/comments/#list-comments-on-an-issue
     * @see https://developer.github.com/v3/#pagination
     *
     * @reamrks
     * Github V3 does not support sort for issue comments now.
     * Github V3 have to request the parent issue to get the count of comments.
     */
    GithubV3.prototype.getComments = function (_a) {
        var accessToken = _a.accessToken, issueId = _a.issueId, _b = _a.query, _c = _b === void 0 ? {} : _b, _d = _c.page, page = _d === void 0 ? 1 : _d, _e = _c.perPage, perPage = _e === void 0 ? 10 : _e, _f = _c.sort, sort = _f === void 0 ? 'desc' : _f;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var issueOptions, commentsOptions, _g, issueRes, commentsRes, linkHeader, thisPage, thisPerPage;
            return tslib_1.__generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        issueOptions = {
                            params: {
                                // to avoid caching
                                timestamp: Date.now(),
                            },
                        };
                        commentsOptions = {
                            params: {
                                // pagination
                                'page': page,
                                'per_page': perPage,
                                /**
                                 * github v3 api does not support sort for issue comments
                                 * have sent feedback to github support
                                 */
                                // 'sort': 'created',
                                // 'direction': sort,
                                // to avoid caching
                                'timestamp': Date.now(),
                            },
                            headers: {
                                'Accept': [
                                    'application/vnd.github.v3.raw+json',
                                    'application/vnd.github.v3.html+json',
                                    'application/vnd.github.squirrel-girl-preview',
                                ],
                            },
                        };
                        if (accessToken) {
                            issueOptions.headers = {
                                'Authorization': "token " + accessToken,
                            };
                            commentsOptions.headers['Authorization'] = "token " + accessToken;
                        }
                        return [4 /*yield*/, Promise.all([
                                this.$http.get("repos/" + this.owner + "/" + this.repo + "/issues/" + issueId, issueOptions),
                                this.$http.get("repos/" + this.owner + "/" + this.repo + "/issues/" + issueId + "/comments", commentsOptions),
                            ])
                            // it's annoying that have to get the page and per_page from the `Link` header
                        ];
                    case 1:
                        _g = tslib_1.__read.apply(void 0, [_h.sent()
                            // it's annoying that have to get the page and per_page from the `Link` header
                            , 2]), issueRes = _g[0], commentsRes = _g[1];
                        linkHeader = commentsRes.headers['link'] || null;
                        thisPage = /rel="next"/.test(linkHeader)
                            ? Number(linkHeader.replace(/^.*[^_]page=(\d*).*rel="next".*$/, '$1')) - 1
                            : /rel="prev"/.test(linkHeader)
                                ? Number(linkHeader.replace(/^.*[^_]page=(\d*).*rel="prev".*$/, '$1')) + 1
                                : 1;
                        thisPerPage = linkHeader ? Number(linkHeader.replace(/^.*per_page=(\d*).*$/, '$1')) : perPage;
                        return [2 /*return*/, {
                                count: Number(issueRes.data.comments),
                                page: thisPage,
                                perPage: thisPerPage,
                                data: commentsRes.data.map(normalizeComment),
                            }];
                }
            });
        });
    };
    /**
     * Create a new comment
     *
     * @param options.accessToken - User access token
     * @param options.issueId - The id of issue
     * @param options.content - The content of comment
     *
     * @return The created comment
     *
     * @see https://developer.github.com/v3/issues/comments/#create-a-comment
     */
    GithubV3.prototype.postComment = function (_a) {
        var accessToken = _a.accessToken, issueId = _a.issueId, content = _a.content;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.post("repos/" + this.owner + "/" + this.repo + "/issues/" + issueId + "/comments", {
                            body: content,
                        }, {
                            headers: {
                                'Authorization': "token " + accessToken,
                                'Accept': [
                                    'application/vnd.github.v3.raw+json',
                                    'application/vnd.github.v3.html+json',
                                    'application/vnd.github.squirrel-girl-preview',
                                ],
                            },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeComment(data)];
                }
            });
        });
    };
    /**
     * Edit a comment
     *
     * @param options.accessToken - User access token
     * @param options.commentId - The id of comment
     * @param options.content - The content of comment
     *
     * @return The edited comment
     *
     * @see https://developer.github.com/v3/issues/comments/#edit-a-comment
     */
    GithubV3.prototype.putComment = function (_a) {
        var accessToken = _a.accessToken, commentId = _a.commentId, content = _a.content;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.patch("repos/" + this.owner + "/" + this.repo + "/issues/comments/" + commentId, {
                            body: content,
                        }, {
                            headers: {
                                'Authorization': "token " + accessToken,
                                'Accept': [
                                    'application/vnd.github.v3.raw+json',
                                    'application/vnd.github.v3.html+json',
                                    'application/vnd.github.squirrel-girl-preview',
                                ],
                            },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeComment(data)];
                }
            });
        });
    };
    /**
     * Delete a comment
     *
     * @param options.accessToken - User access token
     * @param options.commentId - The id of comment
     *
     * @return `true` if succeed, `false` if failed
     *
     * @see https://developer.github.com/v3/issues/comments/#delete-a-comment
     */
    GithubV3.prototype.deleteComment = function (_a) {
        var accessToken = _a.accessToken, commentId = _a.commentId;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var status;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.delete("repos/" + this.owner + "/" + this.repo + "/issues/comments/" + commentId, {
                            headers: { 'Authorization': "token " + accessToken },
                        })];
                    case 1:
                        status = (_b.sent()).status;
                        return [2 /*return*/, status === 204];
                }
            });
        });
    };
    /**
     * Get reactions of a comment
     *
     * @param options.accessToken - User access token
     * @param options.commentId - The id of comment
     *
     * @return The comments
     *
     * @see https://developer.github.com/v3/issues/comments/#get-a-single-comment
     * @see https://developer.github.com/v3/reactions/#list-reactions-for-an-issue-comment
     *
     * @remarks
     * The `List reactions for an issue comment` API also returns author of each reaction.
     * As we only need the count, use the `Get a single comment` API is much simpler.
     */
    GithubV3.prototype.getCommentReactions = function (_a) {
        var accessToken = _a.accessToken, commentId = _a.commentId;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.get("repos/" + this.owner + "/" + this.repo + "/issues/comments/" + commentId, {
                            headers: {
                                'Authorization': "token " + accessToken,
                                'Accept': 'application/vnd.github.squirrel-girl-preview',
                            },
                        })];
                    case 1:
                        data = (_b.sent()).data;
                        return [2 /*return*/, normalizeReactions(data.reactions)];
                }
            });
        });
    };
    /**
     * Create a new reaction of a comment
     *
     * @param options.accessToken - User access token
     * @param options.commentId - The id of comment
     * @param options.reaction - The reaction
     *
     * @return `true` if succeed, `false` if already token
     *
     * @see https://developer.github.com/v3/reactions/#create-reaction-for-an-issue-comment
     */
    GithubV3.prototype.postCommentReaction = function (_a) {
        var accessToken = _a.accessToken, commentId = _a.commentId, reaction = _a.reaction;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var response;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.$http.post("repos/" + this.owner + "/" + this.repo + "/issues/comments/" + commentId + "/reactions", {
                            content: mapReactionName(reaction),
                        }, {
                            headers: {
                                'Authorization': "token " + accessToken,
                                'Accept': 'application/vnd.github.squirrel-girl-preview',
                            },
                        })];
                    case 1:
                        response = _b.sent();
                        return [2 /*return*/, response.status === 201];
                }
            });
        });
    };
    return GithubV3;
}());
export default GithubV3;
